File: Python/OneLang/One/Transforms/ResolveImports.py
from OneLangStdLib import *
import OneLang.One.Ast.Types as types
import OneLang.One.AstTransformer as astTrans

class ResolveImports(astTrans.AstTransformer):
    def __init__(self, workspace):
        self.workspace = workspace
        super().__init__("ResolveImports")
    
    def visit_file(self, source_file):
        ResolveImports.process_file(self.workspace, source_file)
    
    @classmethod
    def process_file(cls, ws, file):
        for imp in file.imports:
            imp_pkg = ws.get_package(imp.export_scope.package_name)
            scope = imp_pkg.get_exported_scope(imp.export_scope.scope_name)
            imp.imports = scope.get_all_exports() if imp.import_all else list(map(lambda x: scope.get_export(x.name) if isinstance(x, types.UnresolvedImport) else x, imp.imports))
            file.add_available_symbols(imp.imports)
    
    @classmethod
    def process_workspace(cls, ws):
        for pkg in ws.packages.values():
            for file in pkg.files.values():
                ResolveImports.process_file(ws, file)

File: PHP/src/OneLang/One/Transforms/ResolveImports.php
<?php

namespace OneLang\One\Transforms\ResolveImports;

use OneLang\One\Ast\Types\Workspace;
use OneLang\One\Ast\Types\UnresolvedImport;
use OneLang\One\Ast\Types\SourceFile;
use OneLang\One\Ast\Types\Package;
use OneLang\One\AstTransformer\AstTransformer;

class ResolveImports extends AstTransformer {
    public $workspace;
    
    function __construct($workspace) {
        parent::__construct("ResolveImports");
        $this->workspace = $workspace;
    }
    
    function visitFile($sourceFile) {
        ResolveImports::processFile($this->workspace, $sourceFile);
    }
    
    static function processFile($ws, $file) {
        foreach ($file->imports as $imp) {
            $impPkg = $ws->getPackage($imp->exportScope->packageName);
            $scope = $impPkg->getExportedScope($imp->exportScope->scopeName);
            $imp->imports = $imp->importAll ? $scope->getAllExports() : array_map(function ($x) use ($scope) { return $x instanceof UnresolvedImport ? $scope->getExport($x->name) : $x; }, $imp->imports);
            $file->addAvailableSymbols($imp->imports);
        }
    }
    
    static function processWorkspace($ws) {
        foreach (array_values($ws->packages) as $pkg)
            foreach (array_values($pkg->files) as $file)
                ResolveImports::processFile($ws, $file);
    }
}
